### PR TITLE
Add metadata attribute to `kedro.io` datasets

### PR DESCRIPTION
**Pull request by AhdraMeraliQB**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Part of #2440 and https://github.com/kedro-org/kedro-plugins/pull/189/

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: kedro/io/cached_dataset.py [hunk 1]
```diff
@@ -50,6 +51,7 @@ def __init__(
             copy_mode: The copy mode used to copy the data. Possible
                 values are: "deepcopy", "copy" and "assign". If not
                 provided, it is inferred based on the data type.
+            metadata: Any arbitrary user metadata.
```
Comment (merelcht) [inline kedro/io/cached_dataset.py:54]: I left a similar comment on `kedro-plugins`, but I'd add a bit more info here about metadata not being used within Kedro.

**Review by merelcht — COMMENTED** [review]
This change should also go into the release notes!

File: kedro/io/lambda_dataset.py [hunk 2]
```diff
@@ -111,3 +115,4 @@ def __init__(
         self.__save = save
         self.__exists = exists
         self.__release = release
+        self.metadata = metadata
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/io/lambda_dataset.py:118]: And of course that would apply to the other datasets as well!

File: kedro/io/partitioned_dataset.py [hunk 3]
```diff
@@ -190,7 +193,8 @@ def __init__(  # pylint: disable=too-many-arguments
         self._filename_suffix = filename_suffix
         self._overwrite = overwrite
         self._protocol = infer_storage_options(self._path)["protocol"]
-        self._partition_cache: Cache = Cache(maxsize=1)
+        self._partition_cache = Cache(maxsize=1)  # type: Cache
```
Comment (merelcht) [inline kedro/io/partitioned_dataset.py:196]: I've recently updated the code base to use a newer way of type annotation: https://github.com/kedro-org/kedro/pull/2538. We should use this syntax going forward. 
```suggestion
        self._partition_cache: Cache = Cache(maxsize=1)
```

**Review by merelcht — APPROVED** [review]
Don't forget to update the release notes! 🙂

File: RELEASE.md [hunk 4]
```diff
@@ -12,6 +12,7 @@
 
 ## Major features and improvements
 * `kedro run --params` now updates interpolated parameters correctly when using `OmegaConfigLoader`.
+* Added `metadata` attribute to kedro.io datasets.
```
Comment (antonymilne) [inline RELEASE.md:15]:
```suggestion
* Added `metadata` attribute to all existing datasets. This is ignored by Kedro, but may be consumed by users or external plugins.
```

**Review by antonymilne — APPROVED** [review]

**Pull request merged**